Refactored SauceOnDemandListener and a minor test update.

mehmetg opened the pull request:
Changes:
* Refactored the listener not to hold on to the SauceSessionIdProvider long term i.e. between the start and end of the test execution). -> The SauceSessionIdProvider was not declared ThreadSafe and therefore was susceptible to corruption.
* Removed SessionId print out from the onTestStart where there's no SessionId to be had. (The session id gets created in test setup or in the test. OnTestStart executes before any of those get executed. Moved it to onTestSuccess and onTestFailure where the information needed is available.
* Changed the title test target to saucelabs test page to make sure we're not at the mercy of amazon for this test to pass.

The reason for the changes:
* While running test methods in parallel on occasion the session id gets not set due to concurrency issues. These changes remedy those issues.
* The old code was making many errors, the changes make it so that unacceptable conditions i.e. no session id being set throw exceptions.

Unit tests ran and passed also tested with sample code and performed as it should.

mehmetg commented:
The local tests are passing consistently. Not sure what went wrong here. There were no changes to the junit module in the PR.

halkeye commented:
@mehmetg can you add a test to highlight what was broken on master, and what is fixed by this change?

mehmetg commented:
@halkeye: Added the tests to cover the changes. 
Also removed the build name update from the test result update method as it was overwriting user set ones. @theCampbell pointed that one out.

halkeye commented on testng/src/test/java/com/saucelabs/testng/SampleSauceTest.java:
```diff
@@ -1,48 +1,32 @@
-import com.saucelabs.common.SauceOnDemandAuthentication;
-import com.saucelabs.common.SauceOnDemandSessionIdProvider;
-import org.openqa.selenium.Platform;
+import org.apache.commons.lang3.tuple.Pair;
```
Can you confirm the imports are all used? I don't see Pair being used in this file.

halkeye replied:
Can you check your imports? I don't think pair is used in this file.

mehmetg replied:
Should have run lint... :) IntelliJ cleaned it up for me. I'll push changes soon.

halkeye commented on testng/src/test/java/com/saucelabs/testng/SampleSauceTest.java:
```diff
@@ -1,48 +1,32 @@
 import org.testng.annotations.Listeners;
 import org.testng.annotations.Test;
 
+import java.io.*;
```
Personal preference, I don't like the *, so I've set the editor to not collapse into wild card imports.

halkeye commented:
The travis tests will fail for pull requests until https://github.com/travis-ci/travis-build/pull/356 gets merged into travis. There's no way to support sauce securely in pull requests till then.

mehmetg commented:
Thanks for reviewing. The local tests have passed, the travis.ci ones are at your discretion. 
I personally think that we should stub those out eventually.

halkeye commented:
:+1: LGTM

I still think it could use more tests.

mehmetg commented:
Agreed. I'll add more in time. Any suggestions on how to test listeners would also be great.
